corehq/apps/app_manager/suite_xml/post_process/workflow.py

```diff
@@ -83,30 +83,47 @@ def get_frame_children(self, command, target_module, module_only=False, include_
         target_module_id = match.group(1)
         target_form_id = match.group(2)
 
-        module_command = id_strings.menu_id(target_module)
+        frame_children = []
+
+        def _add_new_datums(datums):
+            frame_children.extend([d for d in datums if d not in frame_children])
+
+        # Determine datums needed for target form
         module_datums = self.get_module_datums(target_module_id)
         form_datums = module_datums[target_form_id]
-        frame_children = []
 
+        target_module_command = None
+        root_module_command = None
+        root_datums = [[]]
+
+        module_command = id_strings.menu_id(target_module)
         if module_command == id_strings.ROOT:
-            datums_list = self.root_module_datums
+            target_module_datums = self.root_module_datums
         else:
-            datums_list = list(module_datums.values())  # [ [datums for f0], [datums for f1], ...]
+            target_module_command = CommandId(module_command)
+            target_module_datums = list(module_datums.values())  # [ [datums for f0], [datums for f1], ...]
             root_module = target_module.root_module
             if root_module and include_target_root:
-                datums_list = datums_list + list(self.get_module_datums(id_strings.menu_id(root_module)).values())
```

Comment:
This is the main problem: added in https://github.com/dimagi/commcare-hq/pull/16797/files#diff-db3e23918ac72a755c2ca0ecd227d455R88 to fix the ordering of datums in some apps, it was also causing fewer datums to be added in some cases (because only datums in both the root & target would be added to the beginning of the stack).

The main idea of *this* PR is to more painstakingly mimic the user's experience: first add any datums needed by all root menu forms, then add the root menu itself, then any datums needed by all target menu forms, then the target menu itself, then any datums needed by the target form that haven't already been added and the target form itself - which will be popped off, along with possibly datums that don't require user selection, [here](https://github.com/dimagi/commcare-hq/blob/f4a9f630e70a58f568641955ebf53ada76212977/corehq/apps/app_manager/suite_xml/post_process/workflow.py#L288-L294).